Make embedded sso (UserModel::connect) validation more relax for trusted providers

topcheese13 opened the pull request:
Closes #7713 

The problem was with our new validation logic, users with embedded sso couldn't bypass the username validation even when coming from a trusted provider.

The goal here was to apply the same changes to the validation we already did in the EntryController `connect` endpoint. I tried to make it as consistent  to the EntryController as possible in the UserModel `connect` endpoint.

Steps I took to test this:

* Setup up an embedded site (with vanilla-docker)
* Setup up an embedded sso (JS Connect) (with vanilla-docker)
* `Registration.AutoConnect = true`
* I added the vanilla_sso string to my embedded site for every user I was testing (with stub-sso-providers JsSSOString method for JS Connect
* After creating a user and signin with him on the SSO, hitting the homepage of the embedded site was triggering a call to `UserModel->sso()` and `UserModel->connect()`.
* It was successfully registering & signin users
* It was failing to register users when I hard coded the provider to false to force a username validation.

initvector commented on applications/dashboard/models/class.usermodel.php:
```diff
@@ -776,7 +776,7 @@ public function synchUser($currentUser, $newUser, $force = false) {
      * @since 2.1
      */
-    public function syncUser($currentUser, $newUser, $force = false) {
+    public function syncUser($currentUser, $newUser, $force = false, $isTrustedProvider = false) {
```
You probably want to update the doc block to add this new parameter.

initvector commented on applications/dashboard/models/class.usermodel.php:
```diff
@@ -836,9 +841,14 @@ public function connect($uniqueID, $providerKey, $userData, $options = []) {
         trace('UserModel->Connect()');
         $provider = Gdn_AuthenticationProviderModel::getProviderByKey($providerKey);
 
+        $isTrustedProvider = isset($provider['Trusted']) ? $provider['Trusted'] : false;
```
This would be a perfect fit for the [null coalescing operator](http://php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op). I think there are a few other candidates for null-coalescing in here.

If the "Trusted" field of `$provider` isn't already a boolean, I would cast it to one when you're defining `$isTrustedProvider`.

initvector commented on applications/dashboard/models/class.usermodel.php:
```diff
@@ -836,9 +841,14 @@ public function connect($uniqueID, $providerKey, $userData, $options = []) {
+        $saveRoles = $saveRolesRegister = false;
+
         // Trusted providers can sync roles.
-        if (val('Trusted', $provider) && (!empty($userData['Roles']) || !empty($userData['Roles']))) {
```
Neat.

initvector commented on applications/dashboard/models/class.usermodel.php:
```diff
@@ -854,7 +864,7 @@ public function connect($uniqueID, $providerKey, $userData, $options = []) {
         if ($userID) {
             // Save the user.
-            $this->syncUser($userID, $userData);
+            $this->syncUser($userID, $userData, false /*force*/, $isTrustedProvider);
```
Documenting the names of parameters as part of a method call? This looks like a new practice. Is it new? It looks new.

initvector commented on applications/dashboard/models/class.usermodel.php:
```diff
@@ -876,15 +886,16 @@ public function connect($uniqueID, $providerKey, $userData, $options = []) {
+                $options['NoConfirmEmail'] = isset($userData['Email']) || !UserModel::requireConfirmEmail();
+                $options['NoActivity'] = isset($options['NoActivity']) ? $options['NoActivity'] : true;
+                $options['SaveRoles'] = $saveRolesRegister;
+                $options['ValidateName'] = $isTrustedProvider;
```
Is this right? Or do you want to invert the value of `$isTrustedProvider`?

initvector commented:
This looks like it has the potential to be a little hairy, so proper testing is important. Do you have some recommended QA steps you can add to the PR description?

topcheese13 commented:
You are correct with everything you said @initvector.

I updated all the the code related to your remarks and added a bit of testing context in the PR description.

initvector commented on applications/dashboard/models/class.usermodel.php:
```diff
@@ -876,15 +887,16 @@ public function connect($uniqueID, $providerKey, $userData, $options = []) {
                 }
-                touchValue('SaveRoles', $options, !empty($userData['RoleID']) && c('Garden.SSO.SyncRoles', false));
+
+                $options['CheckCaptcha'] = isset($options['CheckCaptcha']) ? $options['CheckCaptcha'] : false;
```
> This would be a perfect fit for the [null coalescing operator](http://php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op). I think there are a few other candidates for null-coalescing in here.

topcheese13 replied:
Think I got all of 'em now.

initvector commented:
> Feel free to assign who you think is the best fit to Q&A this one more time!

I tagged in @cchabilall83 a few days ago, but I know he was out, so I'm mentioning him here, in case he didn't see it!

cchabilall83 commented:
@initvector saw it!  thanks
